grants/search.js

```diff
@@ -1,53 +1,113 @@
 'use strict';
-const request = require('request-promise-native');
 const { head } = require('lodash');
+const request = require('request-promise-native');
 
+/**
+ * 360Giving organisation prefix
+ * We don't want to expose this in public urls
+ * so we prepend this when doing lookups
+ */
 const ID_PREFIX = '360G-blf-';
 
-const isPostcode = input => {
-    // Via https://github.com/chriso/validator.js/blob/master/lib/isPostalCode.js#L54
-    const POSTCODE_PATTERN = /(gir\s?0aa|[a-zA-Z]{1,2}\d[\da-zA-Z]?\s?(\d[a-zA-Z]{2})?)/;
-    return !!input.match(POSTCODE_PATTERN);
+/**
+ * Country regular expressions
+ * geocodes start with a letter prefix denoting the country
+ * we use this to do country lookups.
+ */
+const COUNTRY_REGEXES = {
+    england: /^E/,
+    wales: /^W/,
+    scotland: /^S/,
+    'northern-ireland': /^N/
 };
 
-async function lookupPostcode(postcode) {
-    return request({
-        method: 'GET',
-        url: `https://api.postcodes.io/postcodes/${postcode}`,
-        headers: {
-            'Content-Type': 'application/json'
+/**
+ * Is postcode?
+ * Matches UK postcodes only
+ * @see https://github.com/chriso/validator.js/blob/master/lib/isPostalCode.js#L54
+ * @param {string} input
+ */
+function isPostcode(input) {
+    return /(gir\s?0aa|[a-zA-Z]{1,2}\d[\da-zA-Z]?\s?(\d[a-zA-Z]{2})?)/.test(
+        input
+    );
+}
+
+/**
+ * Build sort criteria
+ * 1. Default to awardDate (newest first)
+ * 2. If we have an explicit sort param use that
+ * 3. Otherwise, if we're performing a text search sort by score
+ * @param {object} queryParams
+ */
+function buildSortCriteria(queryParams) {
+    let sortConf = { awardDate: -1 };
+    if (queryParams.sort) {
+        const [field, direction] = queryParams.sort.split('|');
+        if (['awardDate', 'amountAwarded'].indexOf(field) !== -1) {
+            const newSortConf = {};
+            newSortConf[field] = direction === 'asc' ? 1 : -1;
+            sortConf = newSortConf;
         }
-    }).then(response => JSON.parse(response));
+    } else if (queryParams.q) {
+        sortConf = {
+            score: {
+                $meta: 'textScore'
+            }
+        };
+    }
+
+    return sortConf;
 }
 
+/**
+ * Build $match criteria
+ * @see https://docs.mongodb.com/manual/reference/operator/aggregation/match/index.html
+ * @param {object} queryParams
+ */
 async function buildMatchCriteria(queryParams) {
     const match = {};
 
     match.$and = [];
     match.$or = [];
 
-    // Handle a directly-entered postcode
-    if (queryParams.q && isPostcode(queryParams.q)) {
-        queryParams.postcode = queryParams.q;
-        delete queryParams.q;
+    /**
+     * Default criteria
+     * $facet pipelines must have at least one match in order to
+     * use the index so we set something here that matches everything.
+     */
+    match.$and.push({
```

Comment:
This is the main change that gets `$facet` to use the index. Using `awardDate: { $exists: true }` as a) it will always match and b) it's the field we sort on by default so felt most natural.